### PR TITLE
Batched saving

### PR DESCRIPTION
**Pull request by HarshBabla99**
Instead of updating the h5 file at every epoch, which was quite slow (especially on an HPC cluster), I introduced a `save_period` option to batch saving. 
      - This necessitated some change in `file_io.py`, but I made sure that the final output was the same as before. 
      - The only change is that `parameters` are not saved at every epoch. I assumed that this would be overkill for most use cases anyway. Now, `parameters` are only stored after every `save_period`.

### CODE REVIEW

File: qontrol/optimize.py [hunk 1]
```diff
@@ -155,21 +159,29 @@ def step(
                 else:
                     print(costs, cost_values[0])
 
-            if filepath is not None:
+            # Save and/or plot
+            if filepath is not None and epoch % opt_options['save_period'] == 0 and epoch > 0:
+                if opt_options['verbose']:
+                    print(f'... saving data; from {last_saved_epoch+1} to {epoch}\n')
+                cost_since_last_saved = jnp.asarray(cost_values_over_epochs)[last_saved_epoch+1:]
```
Comment (dkweiss31) [inline qontrol/optimize.py]: I think I'd prefer to only save costs for the epochs satisfying `epoch % opt_options['save_period'] == 0`. This is because a classic technique is finding the epoch of lowest cost with e.g. `argmin`, and using the corresponding parameters. This is a more difficult task if the number of saved costs is not the same as the number of saved parameters

File: qontrol/optimize.py [hunk 2]
```diff
@@ -155,21 +159,29 @@ def step(
...
                 data_dict = {
-                    'cost_values': jnp.asarray(cost_values),
-                    'total_cost': total_cost,
+                    'cost_values': cost_since_last_saved,
+                    'total_cost': jnp.sum(cost_since_last_saved, axis=-1),
```
Comment (dkweiss31) [inline qontrol/optimize.py]: So this would change back to
```suggestion
                    'total_cost': total_cost,
```

File: qontrol/optimize.py [hunk 3]
```diff
@@ -155,21 +159,29 @@ def step(
...
                 }
                 if type(parameters) is dict:
-                    data_dict = data_dict | parameters
+                    for key, val in parameters.items():
+                        data_dict[key] = val[None, ...] # extra axis ensures that they're stacked
```
Comment (dkweiss31) [inline qontrol/optimize.py]: Do we need this? Sometimes I've actually used floats in `parameters`, so this code would fail there

File: qontrol/optimize.py (hunk 1 above)
Comment (HarshBabla99) [reply qontrol/optimize.py]: That makes sense. My main concern is that optimizing within a script and saving the results might not align perfectly with running it in a notebook. If we don't save the cost at each epoch, the saved output wouldn't be able to reproduce the cost-vs-epoch plot from the notebook. 
Furthermore, if no filename is given, running the optimization in a notebook currently only returns the parameters from the final epoch.

How do you think we should resolve these differences?

File: qontrol/optimize.py (hunk 3 above)
Comment (HarshBabla99) [reply qontrol/optimize.py]: I did this after some trial and error; there might be a better way to resolve this. 
Earlier, you were adding the extra axis to the h5 dataset directly (in file_io.py). As a result, all the data would be stacked. My code stacks the parameters, but appends the costs. 

So, suppose the `parameters` has shape `(p1,p2)`, `costs` has shape `(c,)` and we save every `n` epochs. 
Then, after the first save, the saved parameters would have shape `(1,p1,p2)` and the saved costs would have shape `(n,c)`. After `k` saves, the shapes would be `(k,p1,p2)` and `(k*n, c)` respectively. 

The reasons for this distinction are:
- We're saving costs every epoch, but parameters only every `n` epochs. If we implemented the change you suggested above. This discussion would be moot. 
- Suppose we had introduced an extra axis for `costs` as well. In this case, after the first save, the dataset would have the shape `(1,n,c)` and after `k` saves the shape would be `(k,n,c)`. However, we would run into an issue if the optimization is terminated early. In such a case, there would be fewer costs in the unsaved batch (i.e. some `m < n`). In this case, we wouldn't be able to stack `(m,c)` onto `(k,n,c)`. That's why I instead have it such that it does `(k*n, c) -> (k*n + m, c)`. 

You make a good point about scalar parameters. Perhaps a way around this is to convert scalars to a 1D array at the start.

Comment (dkweiss31) [reply qontrol/optimize.py]: Correct me if I'm wrong, but I think this discussion is an artifact of the asymmetry between the way we would save costs and parameters. I firmly believe that the way we save should be symmetric between costs and parameters, to avoid confusion. However, I get that having the costs data (and parameters) from every epoch could be useful even if we save sporadically. My suggestion is to add another option `save_data_from_all_epochs` or something that defaults to `False`. If it is `True`, then we stack the costs and parameters during the intervening periods and can dump them into the h5 when we go to save. If `False`, we only save data at the `save_period`. What do you think?

File: qontrol/optimize.py (hunk 1 above)
Comment (dkweiss31) [reply qontrol/optimize.py]: See below comment

File: qontrol/optimize.py [hunk 4]
```diff
@@ -183,10 +195,30 @@ def step(
...
+        cost_since_last_saved = jnp.asarray(cost_values_over_epochs)[last_saved_epoch+1:]
+        data_dict = {
+            'cost_values': cost_since_last_saved,
+            'total_cost': jnp.sum(cost_since_last_saved, axis=-1),
+        }
+        if type(parameters) is dict:
+            data_dict = data_dict | parameters
+        else:
+            data_dict['parameters'] = parameters
+        append_to_h5(filepath, data_dict, opt_options)
```
Comment (dkweiss31) [inline qontrol/optimize.py]: A lot of code duplication here from above, if we agree we need this extra save then you'd want to make this save code a new function.

I don't think we need this here though: if the code reaches this point for any reason other than KeyboardInterrupt, we will be saving duplicate data. And if we have a KeyboardInterrupt, there's no guarantee that the new version of the data hasn't already been saved. I would kill this code, but happy to chat about it if you disagree :)

File: qontrol/utils/file_io.py [hunk 5]
```diff
@@ -39,13 +39,15 @@ def extract_info_from_h5(filepath: str) -> [dict, dict]:
 def append_to_h5(filepath: str, data_dict: dict, param_dict: dict) -> None:
     with h5py.File(filepath, 'a') as f:
         for key, val in data_dict.items():
+            if val.shape[0] == 0: # takes care of an edge case
+                continue
```
Comment (dkweiss31) [inline qontrol/utils/file_io.py]: What edge case is this? Then there is no data?

File: qontrol/utils/file_io.py [hunk 6]
```diff
@@ -39,13 +39,15 @@ def extract_info_from_h5(filepath: str) -> [dict, dict]:
...
             if key not in f:
                 f.create_dataset(
-                    key, data=[val], chunks=True, maxshape=(None, *val.shape)
+                    key, data=val, chunks=True, maxshape=(None, *val.shape[1:])
                 )
             else:
-                f[key].resize(f[key].shape[0] + 1, axis=0)
-                f[key][-1] = val
+                f[key].resize(f[key].shape[0] + val.shape[0], axis=0)
+                f[key][-val.shape[0]:] = val
```
Comment (dkweiss31) [inline qontrol/utils/file_io.py]: Can you explain why we need these changes?

**Comment by HarshBabla99** [comment]
Here is a summary of the discussion above. I will recommit with the following requested changes: 
- By default, should only save `costs` every `save_period`; to match how `parameters` are stored. This should solve most, if not all, of the smaller comments above. 
- Add another option `save_data_from_all_epochs` that defaults to False. When `True`, then `costs` and `parameters` from all epochs are saved. I'm not sure if we need this, because setting this option to true would be equivalent to `save_period=1`. 
- Create a separate function to log and save all the data, to avoid code duplication. 

Correct me if I missed anything.

**Comment by dkweiss31** [comment]
Yes this is great, thank you! 

I do think there is a use case for `save_data_from_all_epochs=True` since as you mentioned, you get a serious slowdown from the constant file io. However a user could still want to save all cost data and parameters from every epoch. 

With respect to writing the code in a helper function: this may not be necessary, since I argue that the additional save [here](https://github.com/dkweiss31/qontrol/blob/692c51aeed0e38208a3d5aeeffa7b820590e1c31/qontrol/optimize.py#L207-L220) is not necessary. Please let me know if you disagree!

**Comment by HarshBabla99** [comment]
That makes sense! I hadn't appreciated that you meant to save the data from all epochs at the end of the optimization. 
Although, keeping parameters from all epochs might be very memory intensive.

**Comment by dkweiss31** [comment]
I'm actually not super worried about memory: the memory cost of storing the drive parameters over a number of epochs is probably small compared to the memory cost of doing the automatic differentiation (though this would have to be tested)

**Comment by HarshBabla99** [comment]
Since we aren't concerned about storing the parameters over all epochs, I am now just saving once at the end. This makes the code much cleaner, and solves all the issues above. 

Ready for review.

**Comment by dkweiss31** [comment]
Thanks for this Harsh! In reviewing the code, I think I would actually prefer the following solution (very sorry for waffling!): we should always save all data, but save_period should reflect how often we actually do file_io. This would prevent any confusion about which data we are actually saving (its always all data) and would allow the user to not have to always suffer from file_io slowdown, while still doing it often enough if for instance the job is killed midway through, some data is still saved. With the current solution, if the job is killed there isn't a guarantee that data will be saved (the KeyboardInterrupt thing might save us, but it might not).

So to be clear, I think we should be saving data during the loop, at a period reflected by save_period, of all of the data since the last time we saved. It would also make sense to save data outside the loop after everything is completed (or we had a KeyboardInterrupt), since save_period might not be commensurate with the number of epochs. Does this make sense?

**Comment by HarshBabla99** [comment]
No worries, haha :) 

This is a reasonable concern. I thought about this earlier and was considering implementing handling SLURM cancellation in a separate PR. See [https://stackoverflow.com/questions/65326039/how-to-handle-job-cancelation-in-slurm](https://stackoverflow.com/questions/65326039/how-to-handle-job-cancelation-in-slurm). 
However, this would require some effort on the user's end when writing their SLURM script. We could provide a tutorial to help out with that. 

However, I am not entirely sure if this approach would work all the time. Have not tested it too thoroughly. So, it also makes sense to periodically save the data.

File: qontrol/optimize.py [hunk 7]
```diff
@@ -316,7 +316,7 @@ def _terminate_early(
     if dg < opt_options['gtol']:
         termination_key = 1
     # ftol
-    dF = jnp.abs(total_cost - prev_total_cost)
+    dF = np.abs(total_cost - prev_total_cost)
```
Comment (HarshBabla99) [inline qontrol/optimize.py:319]: Why did you change `jnp`->`np`? 
I agree that we shouldn't `jit` this function, since it contains several conditional branches. 
However, executing all operations in JAX should avoid unnecessary GPU->CPU transfer at every epoch.

**Comment by HarshBabla99** [comment]
Thanks, Danny, for wrapping up this commit! 
Everything LGTM except for a couple `jnp` -> `np` changes (mentioned above).

File: qontrol/optimize.py (hunk 7 above)
Comment (dkweiss31) [reply qontrol/optimize.py:319]: We are outside of a jit region here, so as I understand it, it is generally advisable to use numpy instead of jax.numpy in those regions (there is additional overhead to using jax functions that is not necessary if you aren't going to jit).

**Review by dkweiss31 — APPROVED** [review]
Thanks so much for this PR @HarshBabla99 !
